src/firmware/guest/types/snp.rs

```diff
@@ -309,23 +367,263 @@ Launch TCB:
+            _reserved_1: Default::default(),
+            chip_id: [0; 64],
+            committed_tcb: Default::default(),
+            current_build: Default::default(),
+            current_minor: Default::default(),
+            current_major: Default::default(),
+            _reserved_2: Default::default(),
+            committed_build: Default::default(),
+            committed_minor: Default::default(),
+            committed_major: Default::default(),
+            _reserved_3: Default::default(),
+            launch_tcb: Default::default(),
+            _reserved_4: [0; 168],
+            signature: Default::default(),
+        }
+    }
+}
+
+impl Display for AttestationReportV3 {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            r#"
+Attestation Report ({} bytes):
+Version:                      {}
+Guest SVN:                    {}
+{}
+Family ID:                    {}
+Image ID:                     {}
+VMPL:                         {}
+Signature Algorithm:          {}
+Current TCB:
+{}
+{}
+{}
+Report Data:                  {}
+Measurement:                  {}
+Host Data:                    {}
+ID Key Digest:                {}
+Author Key Digest:            {}
+Report ID:                    {}
+Report ID Migration Agent:    {}
+Reported TCB:                 {}
+CPUID Family ID:              {}
+CPUID Model ID:               {}
+CPUID Stepping:               {}
+Chip ID:                      {}
+Committed TCB:
+{}
+Current Build:                {}
+Current Minor:                {}
+Current Major:                {}
+Committed Build:              {}
+Committed Minor:              {}
+Committed Major:              {}
+Launch TCB:
+{}
+{}
+"#,
+            std::mem::size_of_val(self),
+            self.version,
+            self.guest_svn,
+            self.policy,
+            hexdump(&self.family_id),
+            hexdump(&self.image_id),
+            self.vmpl,
+            self.sig_algo,
+            self.current_tcb,
+            self.plat_info,
+            self.key_info,
+            hexdump(&self.report_data),
+            hexdump(&self.measurement),
+            hexdump(&self.host_data),
+            hexdump(&self.id_key_digest),
+            hexdump(&self.author_key_digest),
+            hexdump(&self.report_id),
+            hexdump(&self.report_id_ma),
+            self.reported_tcb,
+            self.cpuid_fam_id,
+            self.cpuid_mod_id,
+            self.cpuid_step,
+            hexdump(&self.chip_id),
+            self.committed_tcb,
+            self.current_build,
+            self.current_minor,
+            self.current_major,
+            self.committed_build,
+            self.committed_minor,
+            self.committed_major,
+            self.launch_tcb,
+            self.signature
+        )
+    }
+}
+
+#[cfg(any(feature = "openssl", feature = "crypto_nossl"))]
+impl Attestable for AttestationReportV3 {
```

Comment:
You should implement these as default behavior in the trait, and then override only when different behavior is required.